### PR TITLE
Translate volume calculations to C++

### PR DESCRIPTION
**Pull request by paulromano**
This PR moves our stochastic volume calculation functionality to C++. The code was actually 95% written by @wbinventor, so huge kudos to him. I simply took what he had, refactored a few pieces, fixed a few bugs, and got it fully working.

### CODE REVIEW

File: src/volume_calc.cpp [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+#ifdef OPENMC_MPI
+      for (int j = 1; j < mpi::n_procs; j++) {
+        int q;
+        MPI_Recv(&q, 1, MPI_INTEGER, j, 0, mpi::intracomm, MPI_STATUS_IGNORE);
+        int buffer[2*q];
+        MPI_Recv(&buffer[0], 2*q, MPI_INTEGER, j, 1, mpi::intracomm, MPI_STATUS_IGNORE);
+        for (int k = 0; k < q; ++k) {
+          for (int m = 0; m < master_indices[i_domain].size(); ++m) {
+            if (buffer[2*k] == master_indices[i_domain][m]) {
+              master_hits[i_domain][m] += buffer[2*k + 1];
```
Comment (smharper) [inline src/volume_calc.cpp]: `break`?

File: src/volume_calc.cpp [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+  int remainder = n_samples_ % mpi::n_procs;
+  int i_start, i_end;
+  if (mpi::rank < remainder) {
+    i_start = (min_samples + 1)*mpi::rank;
+    i_end = i_start + min_samples + 1;
+  }
+  else {
+    i_start = (min_samples + 1)*remainder + (mpi::rank - remainder)*min_samples;
+    i_end = i_start + min_samples;
+  }
```
Comment (smharper) [inline src/volume_calc.cpp]: Double check the math in here.  I tried running a test with 101 samples and 2 procs and I get that `i_end = 101` for the second proc which means only 100 samples will be run.

File: src/volume_calc.cpp [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+
+  // If the material was not previously hit, append an entry to the material
+  // indices and hits lists
+  if (!already_hit) {
+    indices.push_back(i_material);
+    hits.push_back(1);
+  }
+}
+
+std::vector<VolumeCalculation::Result> VolumeCalculation::execute() const
```
[the remark this answers is not in the document]
Comment (paulromano) [reply src/volume_calc.cpp]: Just to be clear, you're suggesting moving `check_hit` further down?

File: src/volume_calc.cpp [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+  if (mpi::rank < remainder) {
+    i_start = (min_samples + 1)*mpi::rank;
+    i_end = i_start + min_samples + 1;
+  }
+  else {
+    i_start = (min_samples + 1)*remainder + (mpi::rank - remainder)*min_samples;
+    i_end = i_start + min_samples;
+  }
+
+#pragma omp parallel
```
[the remark this answers is not in the document]
Comment (paulromano) [reply src/volume_calc.cpp]: Generally our preprocessor directives have not been indented (I see this one is oddly indented so needs to be changed either way). I can indent this and other pragmas, but they will be inconsistent with other directives like `#ifdef`. What do you think?

File: src/volume_calc.cpp [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+      if (!find_cell(&p, false)) continue;
+
+      // TODO: off-by-one
+      int i_material = p.material == MATERIAL_VOID ? p.material : p.material - 1;
+
+      if (domain_type_ == FILTER_MATERIAL) {
+        if (i_material != MATERIAL_VOID) {
+          for (int i_domain = 0; i_domain < n; i_domain++) {
+            if (model::materials[i_material]->id_ == domain_ids_[i_domain]) {
+              this->check_hit(i_material, indices[i_domain], hits[i_domain]);
```
[the remark this answers is not in the document]
Comment (paulromano) [reply src/volume_calc.cpp]: We could, although to be safe we would also need to check that a domain hasn't been specified twice for the same calculation (which is hypothetically possible). Do you want me to implement that?

File: src/volume_calc.cpp [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+        }
+      }
+    }
+#else
+    for (int i_domain = 0; i_domain < n; ++i_domain) {
+      for (int j = 0; j < indices[i_domain].size(); ++j) {
+        master_indices[i_domain].push_back(indices[i_domain][j]);
+        master_hits[i_domain].push_back(hits[i_domain][j]);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (paulromano) [reply src/volume_calc.cpp]: I think you're right: `master_indices = indices` should suffice. I'll give that a shot to make sure it works.

File: src/volume_calc.cpp [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+        MPI_Recv(&buffer[0], 2*q, MPI_INTEGER, j, 1, mpi::intracomm, MPI_STATUS_IGNORE);
+        for (int k = 0; k < q; ++k) {
+          for (int m = 0; m < master_indices[i_domain].size(); ++m) {
+            if (buffer[2*k] == master_indices[i_domain][m]) {
+              master_hits[i_domain][m] += buffer[2*k + 1];
+            }
+          }
+        }
+      }
+#endif
```
[the remark this answers is not in the document]
Comment (paulromano) [reply src/volume_calc.cpp]: I was thinking about that too

File: src/volume_calc.cpp (hunk 2 above)
Comment (paulromano) [reply src/volume_calc.cpp]: I think it's right actually. In that case, proc 1 has samples [0, 51) (51 total) and proc 2 has samples [51, 101) (50 total).

Comment (smharper) [reply src/volume_calc.cpp]: Ah whoops, you're right!  This logic looks good then

File: src/volume_calc.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (smharper) [reply src/volume_calc.cpp]: Fair enough.  I'll leave the choice up to you.

File: src/volume_calc.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (smharper) [reply src/volume_calc.cpp]: Yeah, I know we'd be breaking with precedent here, but I think it's a good idea to start indenting all the preprocessor directives.  It looks like it's a holdover from the strict formatting standards of very old compilers, and I don't see any logical reason why preprocessor stuff should be indented differently from regular code.  No one forces macros to be non-indented, for example.

File: src/volume_calc.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (smharper) [reply src/volume_calc.cpp]: Yup!

File: src/volume_calc.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (paulromano) [reply src/volume_calc.cpp]: Sure, at least for `#pragma` I agree with you, so I'll go ahead and indent these. I do tend to like `#ifdef` being non-indented. It's a nice visual cue that you're literally changing what code is even there. We'll leave that for another day though :)

**Comment by paulromano** [comment]
@smharper Ok, I think everything has been addressed here

File: src/volume_calc.cpp [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+
+namespace model {
+std::vector<VolumeCalculation> volume_calcs;
+}
+
+//==============================================================================
+// VolumeCalculation implementation
+//==============================================================================
+
+VolumeCalculation::VolumeCalculation(pugi::xml_node node) {
```
Comment (smharper) [inline src/volume_calc.cpp]: Missed an opening brace here

File: src/volume_calc.cpp [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+            // If we made it here, the material hasn't yet been added to the master
+            // list, so add entries to the master indices and master hits lists
+            master_indices[i_domain].push_back(indices[i_domain][j]);
+            master_hits[i_domain].push_back(hits[i_domain][j]);
+          }
+        }
+      }
+    }
+#else
+    master_indices = indices;
```
Comment (smharper) [inline src/volume_calc.cpp]: Also need `master_hits = hits`

Comment (paulromano) [reply src/volume_calc.cpp]: Duh...

**Comment by paulromano** [comment]
@smharper Last two comments are now resolved as well

**Comment by paulromano** [comment]
@smharper Need a merge on this if all is good to go.

**Comment by smharper** [comment]
Thanks for the nudge.  Merging now

**Pull request merged**